[autocomplete] Remove tags with the Delete key

sfavello opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

This was to solve the issue that Autocomplete does not handle the functionality of the delete key, more specifically the forward delete key. When creating a Chip and the chip is on hover, it passes on the impression that you can use either the backspace key or the delete key to eliminate the chip. It's best to stay consistent and include the functionality of a delete key. 

https://user-images.githubusercontent.com/90116354/183264769-7e5f870e-bab3-41b6-a86d-fa3935869c9a.mp4

Commit:
Bring accessiblity to autocomplete with the delete key

Commit:
Run yarn prettier

michaldudak commented on packages/mui-base/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -846,6 +846,16 @@ export default function useAutocomplete(props) {
             });
           }
           break;
+        case 'Delete':
+          if (multiple && !readOnly && inputValue === '' && value.length > 0) {
+            const index = focusedTag === -1 ? value.length - 1 : focusedTag;
+            const newValue = value.slice();
+            newValue.splice(index, 1);
+            handleValue(event, newValue, 'removeOption', {
+              option: value[index],
+            });
+          }
+          break;
```
Looks like the code is duplicated. Could you add the `case 'Delete'` to the previous case block?

sfavello replied:
Awesome, thank you! I will do that asap.

michaldudak reviewed (commented):
Good idea, thanks!
It would be great to include this new functionality in tests.

sfavello committed:
Add test for delete and to the backspace switch case for simplicity

sfavello committed:
fix test

sfavello committed:
run prettier

michaldudak commented:
Thanks!

During testing, I found one behavior I'm not sure about. When no tags are selected (the cursor is visible), pressing <kbd>Delete</kbd> removes the last tag (just like the <kbd>Backspace</kbd> key does). This may be a bit confusing, as <kbd>Delete</kbd> usually removes a character to the right of the visible cursor, not to the left as in this case. So, after all, we may need a separate case for `'Delete'` and run the code only if `focusedTag !== -1`.

The behavior for deleting selected tags looks fine.

---

Could you please merge/rebase on the latest master? We've had some issues with uploading regression tests' screenshots in the past, but they have been fixed, and the current master should work fine.

sfavello committed:
change where it wont delete at end of array length but will on hover and it will delete chip to the right

sfavello committed:
format

sfavello commented:
> Thanks!
> 
> During testing, I found one behavior I'm not sure about. When no tags are selected (the cursor is visible), pressing Delete removes the last tag (just like the Backspace key does). This may be a bit confusing, as Delete usually removes a character to the right of the visible cursor, not to the left as in this case. So, after all, we may need a separate case for `'Delete'` and run the code only if `focusedTag !== -1`.
> 
> The behavior for deleting selected tags looks fine.
> 
> Could you please merge/rebase on the latest master? We've had some issues with uploading regression tests' screenshots in the past, but they have been fixed, and the current master should work fine.

`Delete key` will not delete the value at the end but will delete the value to the right when it's on a `focusedTag`. 

**Video update (you can't see but in beginning I was hitting the delete key to show it won't work) 😅 **
_From around 0 to 5 seconds is the test to make sure the delete will only run when `focusedTag !== -1 _`


https://user-images.githubusercontent.com/90116354/196053868-8d4cb460-87a7-4795-8747-680203b7c43e.mov

sfavello committed:
Merge branch 'mui:master' into autocomplete-accessibility-delete

sfavello committed:
fix lint error in test

sfavello committed:
Merge branch 'autocomplete-accessibility-delete' of https://github.com/sfavello/material-uii into autocomplete-accessibility-delete
merge

sfavello committed:
Merge branch 'mui:master' into autocomplete-accessibility-delete

michaldudak commented:
:+1: It does work well now when the cursor is behind tags. But when a tag is focused, <kbd>Delete</kbd> should delete **the focused tag itself**, not the next one.

sfavello committed:
fix where delete key will delete focused tag

sfavello committed:
Merge branch 'autocomplete-accessibility-delete' of https://github.com/sfavello/material-uii into autocomplete-accessibility-delete
update branch

sfavello commented:
I completely misunderstood. 😅 My apologies. I fixed it where it will not delete a chip when the cursor is behind the tags and when on `focusedTag` it will delete the tag that's on focus. I also updated the test to reflect the changes. 😄 

**DEMO**

https://user-images.githubusercontent.com/90116354/198842720-73f8f656-b183-4dc1-b85a-5c757c877cbe.mov

michaldudak commented on packages/mui-base/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -850,6 +850,16 @@ export default function useAutocomplete(props) {
           break;
+        case 'Delete':
+          if (multiple && !readOnly && inputValue === '' && value.length > 0 && focusedTag !== -1) {
+            const index = focusedTag === -1 ? value.length - 1 : focusedTag;
```
No need to check if `focusedTag === -1`, as the outer `if` handles this already.

sfavello replied:
Awesome, thank you! I appreciate it. I went in and made the change. 😄

michaldudak reviewed (commented):
Now the behavior is correct, thanks!
I found one possible improvement to do in the code.

sfavello committed:
erase unnecessary code

michaldudak reviewed (dismissed):
Thanks for your work!

michaldudak reviewed (commented):
There's some issue with tests - would you mind checking it?

sfavello committed:
fix test

sfavello committed:
fix

sfavello committed:
fix

sfavello committed:
testing error

sfavello commented:
So I checked the two failing tests
1. argos states that there is 8 differences detected in the screenshots
2. the e2e test is failing in a `material-docs.spec.ts` where the icon `AcUnit` is not visible. 
I'm still researching around why my changes could have possibly effected these two things, but I'll keep investigating. The other failing test was fixed. 😄

michaldudak commented:
Try merging in the latest master. It should help to solve the problems with Argos and E2E tests.
Could you also add another unit test that would verify if no tags are deleted if they are not highlighted? Something along these lines:

```jsx
it('does not delete any tags with the delete key when no tag is highlighted', () => {
  const handleChange = spy();
  const options = ['one', 'two'];
  render(
    <Autocomplete
      defaultValue={options}
      options={options}
      onChange={handleChange}
      renderInput={(params) => <TextField {...params} autoFocus />}
      multiple
    />,
  );

  const textbox = screen.getByRole('combobox');

  fireEvent.keyDown(textbox, { key: 'Delete' });

  expect(handleChange.callCount).to.equal(0);
  expect(textbox).toHaveFocus();
});
```

sfavello committed:
add test for tags not highlighted

sfavello committed:
fix

sfavello committed:
merge

sfavello committed:
merge main

sfavello commented:
> Try merging in the latest master. It should help to solve the problems with Argos and E2E tests. Could you also add another unit test that would verify if no tags are deleted if they are not highlighted? Something along these lines:
> 
> ```js
> it('does not delete any tags with the delete key when no tag is highlighted', () => {
>   const handleChange = spy();
>   const options = ['one', 'two'];
>   render(
>     <Autocomplete
>       defaultValue={options}
>       options={options}
>       onChange={handleChange}
>       renderInput={(params) => <TextField {...params} autoFocus />}
>       multiple
>     />,
>   );
> 
>   const textbox = screen.getByRole('combobox');
> 
>   fireEvent.keyDown(textbox, { key: 'Delete' });
> 
>   expect(handleChange.callCount).to.equal(0);
>   expect(textbox).toHaveFocus();
> });
> ```

Added test and updated the branch. That fixed the problem. 🥳 . Thank you for your patience. I appreciate it.

michaldudak reviewed (approved):
Looks good to me. Thanks again for your contribution!

oliviertassinari commented:
Oh nice, I didn't think about that <kbd>Del</kbd>. When I was on Windows, I always struggled to reach. this key on my keyboard .
